src/Resolver/MenuResolver.php

```diff
@@ -0,0 +1,59 @@
+<?php declare(strict_types=1);
+
+namespace Torr\MenuBundle\Resolver;
+
+use Torr\MenuBundle\Item\MenuItem;
+use Torr\MenuBundle\Item\ResolvedMenuItem;
+use Torr\MenuBundle\Voter\VoterInterface;
+
+final class MenuResolver
+{
+	public function __construct (
+		/** @var iterable<VoterInterface> $itemVoters */
+		private readonly iterable $itemVoters = [],
+	) {}
+
+	/**
+	 */
+	public function resolveMenu (MenuItem $root) : ResolvedMenuItem
+	{
+		return $this->resolveItem(null, $root);
+	}
+
+	/**
+	 */
+	private function resolveItem (
```

Comment:
How/where would I manipulate a `ResolvedMenu`/`MenuItem`? E.g. when I have a third party integration that wants to hide/remove certain menu entries from a user due to insufficient permissions, wrong locale, etc.?

Reply:
That’s exactly the reason why there is „virtual“. If you have insufficient permissions, you would mark the node as virtual.

Reply:
The only problem we now have is that we can't mark an existing `MenuItem` as virtual, as we'd need to replace the existing `MenuItem` since the `::$virtual` property is readonly. I'd have to know whether a certain node is virtual while creating/building the menu.

From an API/DX standpoint: I'm not too sure I like the naming as it doesn't seem very intuitive. I guess from a developer's point of view, you either want to „disable” or „remove” the `MenuItem` from the tree — either of these actions could set the „virtual” flag internally.

Thoughts?

Reply:
You can’t remove the item from the tree, otherwise the whole selection of active pages can’t work. 

The property is called „virtual“, as it can combine several different concepts under the same abstract term „virtual“.

Reply:
I updated the implementation:

`virtual` was mainly to define whether the item should be rendered (not virtual), or not (virtual). This was a custom field, as "visibility" was defined as "not virtual and has a label".

This was now changed to a flag `visible`, which is always set to `false` if the `label` is not set.